src/Server/TCPHandler.cpp

```diff
@@ -873,17 +909,68 @@ void TCPHandler::receiveQuery()
         client_info.initial_query_id = client_info.current_query_id;
         client_info.initial_address = client_info.current_address;
     }
-    else
-    {
-        query_context->setInitialRowPolicy();
-    }
 
     /// Per query settings are also passed via TCP.
     /// We need to check them before applying due to they can violate the settings constraints.
     auto settings_format = (client_revision >= DBMS_MIN_REVISION_WITH_SETTINGS_SERIALIZED_AS_STRINGS) ? SettingsWriteFormat::STRINGS_WITH_FLAGS
                                                                                                       : SettingsWriteFormat::BINARY;
     Settings passed_settings;
     passed_settings.read(*in, settings_format);
+
+    /// Interserver secret.
+    std::string received_hash;
+    if (client_revision >= DBMS_MIN_REVISION_WITH_INTERSERVER_SECRET)
+    {
+        readStringBinary(received_hash, *in, 32);
+    }
+
+    readVarUInt(stage, *in);
+    state.stage = QueryProcessingStage::Enum(stage);
+
+    readVarUInt(compression, *in);
+    state.compression = static_cast<Protocol::Compression>(compression);
+
+    readStringBinary(state.query, *in);
+
+    /// It is OK to check only when query != INITIAL_QUERY,
+    /// since only in that case the actions will be done.
+    if (!cluster.empty() && client_info.query_kind != ClientInfo::QueryKind::INITIAL_QUERY)
+    {
+#if USE_SSL
+        std::string data(salt);
+        data += cluster_secret;
+        data += state.query;
+        data += state.query_id;
+        data += connection_context.getUserName();
+        data += client_info.initial_user;
+
+        std::string calculated_hash;
+        calculated_hash.resize(32);
+
+        if (received_hash.size() != calculated_hash.size())
+            throw NetException("Unexpected hash received from client", ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT);
+
+        encodeSHA256(data, reinterpret_cast<unsigned char *>(calculated_hash.data()));
+
+        if (calculated_hash != received_hash)
+            throw NetException("Hash mismatch", ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT);
+        /// TODO: change error code?
+
+        query_context->setUserWithoutCheckingPassword(client_info.initial_user, socket().peerAddress());
```

Comment:
If the interserver secret is used `TCPHandler::receiveHello()` should not `Context::setUser()`. Otherwise you're performing calculation of access rights and default settings twice (because `receiveHello()` calls `Context::setUser()`, and `receiveQuery()` calls `Context::setUserWithoutCheckingPassword()`), and still using the cluster user's name and password. Maybe `TCPHandler::receiveHello` can store user name and password until the first query, when you know whether the interserver secret is used.

Reply:
>Maybe TCPHandler::receiveHello() can store user name and password

Some people may consider this as a security breach, but I'm fine with this